docs/en/concepts-and-designs/lal.md

```diff
@@ -0,0 +1,216 @@
+
+<!-- TODO: grok Java library has poor performance, need to benchmark it, the idea is basically the same with `regexp` above -->
+
+### Extractor
+
+Extractors aim to extract metadata from the logs. The metadata can be a service name, a service instance name, an
+endpoint name, or even a trace ID, all of which can be associated with the existing traces and metrics.
+
+- `service`
+
+`service` extracts the service name from the `parsed` result, and set it into the `LogData`, which will be persisted (if
+not dropped) and is used to associate with traces / metrics.
+
+- `instance`
+
+`instance` extracts the service instance name from the `parsed` result, and set it into the `LogData`, which will be
+persisted (if not dropped) and is used to associate with traces / metrics.
+
+- `endpoint`
+
+`endpoint` extracts the service instance name from the `parsed` result, and set it into the `LogData`, which will be
+persisted (if not dropped) and is used to associate with traces / metrics.
+
+- `traceId`
+
+`traceId` extracts the trace ID from the `parsed` result, and set it into the `LogData`, which will be
+persisted (if not dropped) and is used to associate with traces / metrics.
+
+- `segmentId`
+
+`segmentId` extracts the segment ID from the `parsed` result, and set it into the `LogData`, which will be
+persisted (if not dropped) and is used to associate with traces / metrics.
+
+- `spanId`
+
+`spanId` extracts the span ID from the `parsed` result, and set it into the `LogData`, which will be
+persisted (if not dropped) and is used to associate with traces / metrics.
+
+- `metrics`
+
+`metrics` extracts / generates metrics from the logs. The supported metrics are `CounterMetrics`, `GaugeMetrics`,
+and `HistogramMetrics`, examples are as follows:
+
+```groovy
+filter {
+    // ...
+    extractor {
+        service parsed["serviceName"]
+        metrics {
+            counter { // whenever the filter is fed with a piece of log, the counter will be increased by one
+                name: "logsCount"
+                tips: "The total count of received logs"
+                tags: ["key1": "value1", "key2": "value2"]
+            }
+            if (parsed["level"] == "ERROR") {
+                counter {
+                    // whenever the filter is fed with a piece of log whose level is ERROR, the counter will be increased by one
+                    name: "errorLogsCount"
+                    tips: "The total count of error logs"
+                    tags:
+                    ["key1": "value1", "key2": "value2"]
+                }
+            }
+            gauge { // whenever the filter is fed with a piece of log, the gauge will be increased by one
+                name: "whatever"
+                tips: "whatever"
+                tags: ["k1": "v1", "k2": "v2"]
+            }
+            histogram {
+                name: "httpRequestTime"
+                tips: "The time elapsed of the http request"
+                startNanos: parsed["timestamp"]
```

Comment:
`timestamp` is required by `Sample`. And timestamp is required by all types of metrics.

All metrics generated in one LAL scripts, should be as a SampleFamily, to be the input of the MAL core. Then, users should be able to set the MAL script to do further analysis. Then closed loop is set.

If MAL is still not clear to you, ping me.